### PR TITLE
feat: GraphQL commit query parameters

### LINKED ISSUE #2838: Support GraphQL filter parameters in `commits`

### PR DESCRIPTION
**Pull request by peasee**
## 🗣 Description

* Adds support for using the `since` and `until` query parameters in the GraphQL query for commits
* In the process of adding support, refactors some of the functions to support reusing them more generically
* Updates the GraphQL AST field finder to properly recurse nested fields to find the target field

Enables filter push down on any query mode when searching on `committed_date` like:

```bash
sql> select sha from spiceai.commits where committed_date < '2024-09-24' and committed_date
 > '2024-09-14'
+------------------------------------------+
| sha                                      |
+------------------------------------------+
| 44917e212e6392d75fbb1742e22368af89fe3f81 |
| 9c862a78376e01dcf515f9a699eb01353eed3b23 |
| 269e8d9ee028e011a22bc0ab0f4db0ff39b55b6c |
| 269aad2798c77585e2c4deb486ddbe62357200c7 |
| 36f38ea34fde947854cb9bd84f3912dac1ae4b7b |
| d75b9f88712594a5b8fd0e4dad23afeb24dfcc4b |
| 72fed656586dd71d44b7f139fea0caaf14647da0 |
| fca3ab602b1e1e8d11e3dab5a6348eceb03fd893 |
| 3f985d3714228467e9db7434e48ee491edb49d31 |
| 11f79e33b5321a8a86a048cde4212635147f0808 |
| c86b7ff5e581aa6fc3183cbeb074156b7c13993d |
| 678176dc2e004f8f5c117df31aa6c2e309e20bbb |
| 8851f2d9ede8a2d4964fb57b2a7d49b3cc73a1ca |
| aceda51de62d045fe96df600a1082ef256441e1f |
| 6723fe8307b282ad519c5225fb27528394c6db0e |
| 5e776741e27151f3d2a604ccd2a3f3036796dbdd |
| bcfe39f81bbc4c947035e8db1b130e4b15371222 |
| e70249c67e7c8cc71e2ed439a5b0079faaa657a0 |
| f3285f3d4f3c54dc5b5683dfc7a0bddf1e4b2826 |
| e68728e6b0ec3c4b69846a3487cc69b289669797 |
| 23a247ac8b7f31104aba1fdcdb94d0aa1d4ff382 |
| 793f083e3ca6baafbb8ea699802c55a190f4b2ae |
| 463438cb3758109a84869ffda3ae8716be08963a |
| 700830157ae3d8e0e35c30c4edf29173abd8a810 |
| a49c430adae829234b182f9488055b442389ec27 |
| 145e696619ae01d34ba321fc9f33bc061b358de3 |
| bb6fa845ba1327a6ea9c0e280fc8f018c2601e27 |
| 5cd8f2521b8a48f0bf736fb7ae92f3293e83fa25 |
| 3250850d42bacd98f830bb53630bf75598427baf |
| 7874bb81b48550d4472fef1a213dc341770b9c15 |
| 268689ca09ebcdd664b8ca5a1f72bf02f4859fb8 |
| 674b39fbb1d94d1550b2d0b0725caf98b2b38cd5 |
| 5b82cbd45176244df9fe7100b9de47030f80d57c |
| 617f0b588ea77d8a0b195ee74fa1be7618b5edfb |
| a0d5587ec5146f020190e9c6685d76b5d0d05170 |
| 73a680089f2822a5a6a4cba74abe9adbcb232ff4 |
| 365c8e3bb880e0783365b44d42215162f3556743 |
| f6bd8d4f0ab5a660f6ee875f1584166d5de17a9a |
| 30254fbdc39e79ad108400b9e2ec1c1030648787 |
| e57680199b97a9f3a91e7819bbca4b16f7740efb |
| e340d19fdbef0f5b249a6ad6cc4465af3b3dc3f8 |
| 8f33c309063168795d6a949149a66c8d3a9d2f2b |
| 25417b9c5011dade1f7c74d85f0725b2ec085e45 |
| 61d8e1b9512deca1f2c38662ba2caf28480c5b03 |
| cdce71a475607ab60d0b6caadbd147f330069dea |
| a0e1e881e5b9cba3c08e93fc8f48d1c2a8dfb6a8 |
| 5e1fc3e48309d6aa7cdd3ad9c15dda027abf528d |
| 5358de18bccd1145c1cf00f8be61fdc4ec96fe4d |
| 68cad8e53d529b77416f660b4ea0e18b548db46b |
| 5cfeca37c7e69b1e86f58d0eb8158b48a28492c1 |
| 005215a1d6214944a2a3b317d5d16d4a3f05c4fd |
| 32768b9ffa58fb7927122b638f0597f7dab38dd4 |
| 1bd40eed716d7e11e1d0ee59d78bcd103b68687b |
| a5833fe96803b9e7113c83851d8a68faacbd4bd0 |
| 2e9e766e94b70af914b0b5ed6156f579f0ed6e1f |
| 7b99ca1e1124c82f3920dadf76dace3a9c5c2f11 |
| fa176b086bf171666ae7cb453ae87b00be150ca1 |
| 6134f086c40448b4f7eb742a4e165560395b36f9 |
| 4e11257698e1ecaba3677e14f0ab8664f0677e77 |
| 08dd47cfb00cf4435f37898a631f3a298677f45c |
| ccd722f5d30d13c0bf4bea494acb897db769c8f9 |
| 1b7e8906f74fe5740a1a3f4494426152844456d7 |
| c5dcb108f3ec5513e8010e9a46db1338e4fef87e |
| 61720d9e2401b0d632ab18142b5b2f7f8b1ebabe |
| 95f53d7d3f12bd5bb6faf5850d5882037da71fc2 |
| a3d64982b6707fa1b18593319f8b0b66d0d0d880 |
| f148d43e291f23249fbd8746979bba334809c653 |
| 7d6ca635c2865daafb0c8209e1081f888aa35507 |
| 8321d118eff75373aa52ff0399e3dddb51e5347a |
| 954b63d0422e9acaa4f6ff0c033569b5091b08e0 |
| de48742dc8ae3157f98dfbe49bd8cdd8d5d80655 |
| 6911ce98cc869fda4b9fc2bfc2eac02c3bd4710e |
| 7e8149e9cca6f2627aab509b55611f079b9c7099 |
| d2a3a6e001134f3135ab7db94055d5b9dc47d222 |
| 06d97d7f4d6aacea1079304f50b8c27ffc5285a3 |
+------------------------------------------+

Time: 0.918523286 seconds. 75 rows.
```

Query mode currently has no effect for commits, as it would be implemented in #2835

## 🔨 Related Issues

* Closes #2838

### CODE REVIEW

File: crates/runtime/src/dataconnector/github/mod.rs [hunk 1]
```diff
@@ -510,14 +529,35 @@ lazy_static! {
...
                     Operator::GtEq,
                 ],
-                remap: Some("merged"),
+                remaps: Some(vec![GitHubFilterRemap::Column("merged")]),
+                uses_modifiers: true
+            },
+        );
+
+        m.insert(
+            "committed_date",
```
Comment (eadgbear) [inline crates/runtime/src/dataconnector/github/mod.rs:542]: Can you create an issue to update the Github docs to cover this? The `Spec` section from your Issue covers a good chunk of what we'd need

**Review by eadgbear — DISMISSED** [review]

File: crates/runtime/src/dataconnector/github/mod.rs [hunk 2]
```diff
@@ -404,9 +410,15 @@ pub fn parse_globs(input: &str) -> super::DataConnectorResult<Arc<GlobSet>> {
...
+}
+
 struct GitHubPushdownSupport {
     ops: Vec<Operator>,
-    remap: Option<&'static str>,
+    remaps: Option<Vec<GitHubFilterRemap>>,
+    uses_modifiers: bool,
 }
 
 // TODO: add support for LIKE and IN filters, to support columns like assignees, labels, etc
```
Comment (phillipleblanc) [inline crates/runtime/src/dataconnector/github/mod.rs]: Do these TODO comments need to be updated?

File: crates/runtime/src/dataconnector/github/mod.rs [hunk 3]
```diff
@@ -404,9 +410,15 @@ pub fn parse_globs(input: &str) -> super::DataConnectorResult<Arc<GlobSet>> {
...
+enum GitHubFilterRemap {
+    Column(&'static str),
+    Operator((Operator, &'static str)),
+}
+
 struct GitHubPushdownSupport {
     ops: Vec<Operator>,
-    remap: Option<&'static str>,
+    remaps: Option<Vec<GitHubFilterRemap>>,
+    uses_modifiers: bool,
```
Comment (phillipleblanc) [inline crates/runtime/src/dataconnector/github/mod.rs:425]: What does `uses_modifiers` mean?

File: crates/runtime/src/dataconnector/github/mod.rs (hunk 1 above)
Comment (peasee) [reply crates/runtime/src/dataconnector/github/mod.rs:542]: https://github.com/spiceai/docs/issues/435

File: crates/runtime/src/dataconnector/github/mod.rs (hunk 3 above)
Comment (peasee) [reply crates/runtime/src/dataconnector/github/mod.rs:425]: Whether the query parameter supports e.g. `<` or `>` in the query string. I've added some comments inline as well.

**Review by phillipleblanc — APPROVED** [review]

**Pull request merged**